Unsubdivide and bucket fixes.

jaynabonne opened the pull request.

jaynabonne committed:
Fixes for subdivide and for adding to right "bucket".

sairam4123 commented on addons/sairam.quadtree/QuadTree.gd:
```diff
@@ -195,14 +216,13 @@ func _get_child(body_bounds: AABB) -> QuadTree:
 	"""
-	if body_bounds.end.z < _bounds.end.z:
-		if body_bounds.end.x < _bounds.end.x:
+	if body_bounds.end.z < _bounds.position.z + _center.z:
```
Shouldn't it be `end`?

jaynabonne replied:
You want to go up/down and left/right based on splitting the bounds in half. If it's less than half, go down/up, etc.

sairam4123 commented on addons/sairam.quadtree/QuadTree.gd:
```diff
@@ -129,7 +140,7 @@ func clear() -> void:
 		for child in _children:
 			child.clear()  # clear all it's children
-		_is_leaf = true
+		_set_as_empty_leaf()
```
Why was this moved to a function?

jaynabonne replied:
It needed to be done in three places, and what was done wasn't trivial - it needed to both set is_leaf *as well as* set all children to null. It seemed to be enough of a concept to warrant its own function. It's more than just setting is_leaf now. That was part of the unsubdivide problem - _children wasn't being set back to all null.

sairam4123 committed:
Merge branch 'development' into unsubdivide-fixes

sairam4123 committed:
Update QuadTree.gd

sairam4123 commented on addons/sairam.quadtree/QuadTree.gd:
```diff
@@ -48,35 +53,42 @@ func add_body(body: VisualInstance) -> Object:
 		_subdivide()
 		# update the body's quadtree.
-		return update_body(body)
+#		return update_body(body, false)
```
There was an issue so I just disabled this line.

jaynabonne replied:
Let me look. Without that, it would hang for me.

sairam4123 replied:
Well, don't forget to update your code.

jaynabonne committed:
- Subdivide now moves local objects to children
- Fixed return types for various methods
- Got rid of the "do_unsubdivide" parameters, as well as the need for them.
- Fixed/simplified update_body.
- Got rid of _remove_duplicates, since it should never be needed.
- Got rid of _found_objects member variable, since it's not permenant state.
- Got rid of null return in get_child, since it can't be reached.

jaynabonne committed:
Merge branch 'unsubdivide-fixes' of https://github.com/sairam4123/QuadTree-Godot into unsubdivide-fixes

# Conflicts:
#	addons/sairam.quadtree/QuadTree.gd

jaynabonne committed:
Fix infinite recursing dump. Make logging consistent.

jaynabonne committed:
Fix dump directory creation.

sairam4123 reviewed (commented):
Looks fine for me, except Comments.
Please update the comments and add docstrings.

jaynabonne committed:
- Add function doc comments.
- Fix query returning the same object multiple times. (It only wanted to use one method.)
- Change child ownership test to use center of object, not bottom right corner. (Either would probably work.)
